### PR TITLE
make agent support for ddpg & sac

### PR DESCRIPTION
**Pull request by witwolf**
1.  Modify `Agent` to support  for ddpg, sac .

2. remove `with_experience` from func `rollout`, make it easier to understand 
This only affect `_reward_normalizer` in ICM (may calculate intrinsic  reward multiple times ) , and i have tested on `ppo_icm_super_mario_intrinsic_only` ,it does not affect performance (right:master, left:PR_agent1)

TODO:
since we can train  an `OnPolicyAlgorithm`  with an `OffPolicyTrainer` , it's no necessary to calculate all things (loss, info ...) when rollout , it can save much computation . what to calculate is actually determined by what trainer is used .

### CODE REVIEW

**Comment by emailweixu** [comment]
Sac/ddpg cannot work with icm even with this change, because icm need to have use_rollout_state=False when using off_policy_driver. Perhaps we can leave that as the next PR. As to removing with_experience, @hnyu may not agree because he added that to save the computation for his own intrinsic reward algorithm.

**Comment by hnyu** [comment]
> 1. Modify `Agent` to support  for ddpg, sac .
> 2. remove `with_experience` from func `rollout`, make it easier to understand
>    This only affect `_reward_normalizer` in ICM (may calculate intrinsic  reward multiple times ) , and i have tested on `ppo_icm_super_mario_intrinsic_only` ,it does not affect performance (right:master, left:PR_agent1)
> 
> 
> 
> TODO:
> since we can train an `OnPolicyAlgorithm` with an `OffPolicyTrainer` , it's no necessary to calculate all things (loss, info ...) when rollout , it can save much computation . what to calculate is actually determined by what trainer is used .



> Sac/ddpg cannot work with icm even with this change, because icm need to have use_rollout_state=False when using off_policy_driver. Perhaps we can leave that as the next PR. As to removing with_experience, @hnyu may not agree because he added that to save the computation for his own intrinsic reward algorithm.

Right, with_experience is mainly for saving computational burden. And I think it will serve a general purpose of differentiating quantities that should be computed once during exp collection from quantities that can be computed&optimized multiples times (e.g., costs) during one training iteration.

**Comment by hnyu** [comment]
> Sac/ddpg cannot work with icm even with this change, because icm need to have use_rollout_state=False when using off_policy_driver. Perhaps we can leave that as the next PR. As to removing with_experience, @hnyu may not agree because he added that to save the computation for his own intrinsic reward algorithm.

Why do you say that "icm need to have use_rollout_state=False when using off_policy_driver" ?

**Comment by hnyu** [comment]
> Sac/ddpg cannot work with icm even with this change, because icm need to have use_rollout_state=False when using off_policy_driver. Perhaps we can leave that as the next PR. As to removing with_experience, @hnyu may not agree because he added that to save the computation for his own intrinsic reward algorithm.

Why do you say that "icm need to have use_rollout_state=False when using off_policy_driver" ?

**Comment by hnyu** [comment]
>    This only affect `_reward_normalizer` in ICM (may calculate intrinsic  reward multiple times ) , and i have tested on `ppo_icm_super_mario_intrinsic_only` ,it does not affect performance (right:master, left:PR_agent1)
> 
> 


Seems to me that the left figure has a better sample efficiency than the right one. Is that just because of initialization randomness?

File: alf/algorithms/sac_algorithm.py [hunk 1]
```diff
@@ -349,14 +349,8 @@ def train_step(self, exp: Experience, state: SacState):
         info = SacInfo(actor=actor_info, critic=critic_info, alpha=alpha_info)
         return PolicyStep(action_distribution, state, info)
 
-    def train_complete(self,
-                       tape: tf.GradientTape,
-                       training_info: TrainingInfo,
-                       weight=1.0):
-        ret = super().train_complete(
-            tape=tape, training_info=training_info, weight=weight)
+    def after_train(self):
```
Comment (emailweixu) [inline alf/algorithms/sac_algorithm.py]: need to implement the following (see Algorithm._trainable_attributes_to_ignore)
```python
def _trainable_attributes_to_ignore(self):
    return ['_target_critic_network1', '_target_critic_network2']
```

Comment (witwolf) [reply alf/algorithms/sac_algorithm.py]: added

**Comment by witwolf** [comment]
> 
> Right, `with_experience` is mainly for saving computational burdens and for normalizers&averagers to have correct statistics. And I think it will serve a general purpose of differentiating quantities that should be computed once during exp collection from quantities that can be computed&optimized multiples times (e.g., costs) during one training iteration.

set `calc_intrinsic_reward=False` when collecting experiences in the new commit

File: alf/algorithms/agent.py [hunk 2]
```diff
@@ -141,20 +141,28 @@ def predict(self, time_step: ActionTimeStep, state: AgentState):
...
+        return PolicyStep(action=rl_step.action, state=new_state, info=())
+
+    def rollout(self, time_step: ActionTimeStep, state: AgentState):
         """Rollout for one step."""
         new_state = AgentState()
         info = AgentInfo()
         observation = self._encode(time_step)
 
+        # TODO
+        # avoid computing this when rl_algorithm do not need preprocess experience
```
Comment (hnyu) [inline alf/algorithms/agent.py]: What do you mean by this comment? I believe that for any algorithm, icm rewards should always be computed during rollout.

File: alf/algorithms/agent.py [hunk 3]
```diff
@@ -164,7 +172,32 @@ def rollout(self,
...
+        # TODO
+        # avoid computing this when collecting exps (off policy train)
+        if self._entropy_target_algorithm:
+            et_step = self._entropy_target_algorithm.train_step(rl_step.action)
+            info = info._replace(entropy_target=et_step.info)
+
+        return PolicyStep(action=rl_step.action, state=new_state, info=info)
+
+    def train_step(self, exp: Experience, state):
+        new_state = AgentState()
```
Comment (hnyu) [inline alf/algorithms/agent.py]: Do you need to first convert Experience to ActionTimeStep here?

File: alf/algorithms/rnd_algorithm.py [hunk 4]
```diff
@@ -98,8 +100,11 @@ def train_step(self, inputs, state):
             tf.square(pred_embedding - tf.stop_gradient(target_embedding)),
             axis=-1)
 
-        intrinsic_reward = tf.stop_gradient(loss)
-        intrinsic_reward = self._reward_normalizer.normalize(intrinsic_reward)
+        intrinsic_reward = ()
+        if calc_intrinsic_reward:
+            intrinsic_reward = tf.stop_gradient(loss)
+            intrinsic_reward = self._reward_normalizer.normalize(
+                intrinsic_reward)
```
Comment (hnyu) [inline alf/algorithms/rnd_algorithm.py]: Please also add this check in ICMAlgorithm

File: alf/algorithms/sac_algorithm.py [hunk 5]
```diff
@@ -380,3 +380,6 @@ def _calc_critic_loss(self, training_info):
 
         critic_loss = critic_loss1.loss + critic_loss2.loss
         return LossInfo(loss=critic_loss, extra=critic_loss)
+
+    def _trainable_attributes_to_ignore(self):
+        return ['target_critic_network1', 'target_critic_network2']
```
Comment (emailweixu) [inline alf/algorithms/sac_algorithm.py]: should be ['_target_critic_network1", '_target_critic_network2']

Comment (witwolf) [reply alf/algorithms/sac_algorithm.py]: fixed

File: alf/algorithms/agent.py (hunk 3 above)
Comment (witwolf) [reply alf/algorithms/agent.py]: It's not necessary , and _rl_algorithm.train_step require Experience as input

File: alf/algorithms/agent.py (hunk 2 above)
Comment (witwolf) [reply alf/algorithms/agent.py]: Removed,  I thought it should be  calculated  at  `train_step` for some off policy algorithms that do not need preprocessing experiences
